### ISSUE TITLE
Incorrect handling of residue pairs with zero score at the start of local alignment

### ISSUE THREAD

**Issue by tanhevg** [biopython/biopython#1913]
I would expect the snippet below to print more alignments, that have residue pairs ('Q', 'N') and ('N', 'E') prepended to the aligned region (see output from my own implementation). BLOSUM62 scores for ('Q', 'N') and ('N', 'E') are set to zero, so prepending these residues to the alignment would keep the score at 21. Note that there is a separate alignment with ('D', 'L') (which also scores zero) appended at the __end__ of the aligned region, so the problem is only there with zero-scoring pairs at the beginning of the alignment. Both  `Bio.pairwise2` and `Bio.Align` have this issue.

My environment is printed at the bottom of this post.

```python 
from Bio import Align
from Bio.SubsMat.MatrixInfo import blosum62
from Bio.pairwise2 import align, format_alignment

aligner = Align.PairwiseAligner()

aligner.substitution_matrix = blosum62
aligner.mode = 'local'
aligner.open_gap_score = -11
aligner.extend_gap_score = -1

str1 = 'ISLKYTAARLHEKGVLLEIEDLQVNQFKNVIFEISPTEEVGDFEVKAKFMGVQMETFMLHYQDLLQLQYEGVAVMKLFDRAKVNVNLLIFLLNKKFYGK'
str2 = 'MRIGIISVGPGNIMNLYRGVKRASENFEDVSIELVESPRNDLYDLLFIPGVGHFGEGMRRLRENDLIDFVRKHVEDERYVVGVCLGMQLLFEESEEAPGVKGLSLIEGNVVKLRSRRLPHMGWNEVIFKDTFPNGYYYFVHTYRAVCEEEHVLGTTEYDGEIFPSAVRKGRILGFQFHPEKSSKIGRKLLEKVIECSLSRR'

alignments = aligner.align(str1, str2)
for al in alignments:
    print(al)

loc_alignments = align.localds(str1, str2, blosum62, -11, -1)
for alignment in loc_alignments:
    print(format_alignment(*alignment))

# My own implementation
from alignments import SmithWatermanAligner
aligner = SmithWatermanAligner(11, 1, blosum62)
alignment = aligner.align(str1, str2)
print(alignment.score, alignment, sep='\n')

"""
ISLKYTAARLHEKGVLLEIEDLQVNQFKNVIFEI--SPTEEVGDFEVKAKFMGVQMETFMLHYQDLLQLQYEGVAVMKLFDRAKVNVNLLIFLLNKKFYGK....................................................................................................
..........................|XX|XX|X--||XXXXX|.............................................................................................................................................................
MRIGIISVGPGNIMNLYRGVKRASENFEDVSIELVESPRNDLYDLLFIPGVGHFGEGMRRLRENDLIDFVRKHVEDERYVVGVCLGMQLLFEESEEAPGVKGLSLIEGNVVKLRSRRLPHMGWNEVIFKDTFPNGYYYFVHTYRAVCEEEHVLGTTEYDGEIFPSAVRKGRILGFQFHPEKSSKIGRKLLEKVIECSLSRR

ISLKYTAARLHEKGVLLEIEDLQVNQFKNVIFEI--SPTEEVGDFEVKAKFMGVQMETFMLHYQDLLQLQYEGVAVMKLFDRAKVNVNLLIFLLNKKFYGK....................................................................................................
..........................|XX|XX|X--||XXXXX|X............................................................................................................................................................
MRIGIISVGPGNIMNLYRGVKRASENFEDVSIELVESPRNDLYDLLFIPGVGHFGEGMRRLRENDLIDFVRKHVEDERYVVGVCLGMQLLFEESEEAPGVKGLSLIEGNVVKLRSRRLPHMGWNEVIFKDTFPNGYYYFVHTYRAVCEEEHVLGTTEYDGEIFPSAVRKGRILGFQFHPEKSSKIGRKLLEKVIECSLSRR

ISLKYTAARLHEKGVLLEIEDLQVNQFKNVIFEI--SPTEEVGDFEVKAKFMGVQMETFMLHYQDLLQLQYEGVAVMKLFDRAKVNVNLLIFLLNKKFYGK----------------------------------------------------------------------------------------------------
                          |..|..|.  ||.....|.
MRIGIISVGPGNIMNLYRGVKRASENFEDVSIELVESPRNDLYDLLFIPGVGHFGEGMRRLRENDLIDFVRKHVEDERYVVGVCLGMQLLFEESEEAPGVKGLSLIEGNVVKLRSRRLPHMGWNEVIFKDTFPNGYYYFVHTYRAVCEEEHVLGTTEYDGEIFPSAVRKGRILGFQFHPEKSSKIGRKLLEKVIECSLSRR
  Score=21

ISLKYTAARLHEKGVLLEIEDLQVNQFKNVIFEI--SPTEEVGDFEVKAKFMGVQMETFMLHYQDLLQLQYEGVAVMKLFDRAKVNVNLLIFLLNKKFYGK----------------------------------------------------------------------------------------------------
                          |..|..|.  ||.....|
MRIGIISVGPGNIMNLYRGVKRASENFEDVSIELVESPRNDLYDLLFIPGVGHFGEGMRRLRENDLIDFVRKHVEDERYVVGVCLGMQLLFEESEEAPGVKGLSLIEGNVVKLRSRRLPHMGWNEVIFKDTFPNGYYYFVHTYRAVCEEEHVLGTTEYDGEIFPSAVRKGRILGFQFHPEKSSKIGRKLLEKVIECSLSRR
  Score=21

21.0
NQFKNVIFEI--SPTEEVGDF
++|++|.+|+  ||.+++.|+
ENFEDVSIELVESPRNDLYDL
"""
```

```python
import sys; print(sys.version)
import platform; print(platform.python_implementation()); print(platform.platform())
import Bio; print(Bio.__version__)
"""
3.7.1 | packaged by conda-forge | (default, Nov 13 2018, 18:33:04) 
[GCC 7.3.0]
CPython
Linux-3.10.0-862.14.4.el7.x86_64-x86_64-with-centos-7.5.1804-Core
1.73
"""
```

**Comment by mdehoon** [comment]
Is there some reference in the literature (textbook or authoritative paper) that specifies how this should be handled? It seems strange to extend an alignment if it does not increase the score.

**Comment by tanhevg** [comment]
I would expect **all** alignments with the same score to be returned

**Comment by peterjc** [comment]
From an end user perspective, I can certainly see a desire to have the longer (and equally scoring) alignments returned.

CC @MarkusPiotrowski

**Comment by mdehoon** [comment]
But you wouldn't return a local alignment that has zero score. Can you compare with other software packages (for example Ensembl EMBOSS Water) to see how they handle this case?

**Comment by tanhevg** [comment]
> But you wouldn't return a local alignment that has zero score. 

Why not? If zero is the highest score that can be reached by aligning these sequences,  why not return the alignment that attains it?

**Comment by mdehoon** [comment]
If we allow local alignments with a zero score, then alignments consisting of gaps only would be valid alignments.

EMBOSS water gives the short alignment only:
```
                  27 FKNVIFEI--SPTEEVGD     42
                     |::|..|:  ||..::.|
                  27 FEDVSIELVESPRNDLYD     44
```

**Comment by tanhevg** [comment]
> If we allow local alignments with a zero score, then alignments consisting of gaps only would be valid alignment

Only if gap penalty is zero,  which might be the case for some DNA alignments. The common setup for aligning protein sequences is using BLOSUM62 substitution marix, and gap opening penalty of -11, and extension penalty of -1. Some off diagonal entries of BLOSUM62 are 0, therefore alignments with zero score are legit.

**Comment by mdehoon** [comment]
In the chapter on Smith-Waterman alignments in "Biological Sequence Analysis" by Richard Durbin et al., they write "The traceback ends when we meet a cell with value 0, which corresponds to the start of the alignment."

I am open to changing the implementation such that at the end, no alignments are added with zero score, to be consistent in how the local alignment is defined at the beginning and at the end of the alignment. For your example, the aligner will then only return the alignment found by EMBOSS water.

To get the behavior you are looking for, you can always add some small number to the scoring matrix. The aligner will then extend the alignment to include the pairs with zero score:
```python
>>> from Bio import Align
>>> from Bio.SubsMat.MatrixInfo import blosum62
>>> str1 = 'ISLKYTAARLHEKGVLLEIEDLQVNQFKNVIFEISPTEEVGDFEVKAKFMGVQMETFMLHYQDLLQLQYEGVAVMKLFDRAKVNVNLLIFLLNKKFYGK'
>>> str2 = 'MRIGIISVGPGNIMNLYRGVKRASENFEDVSIELVESPRNDLYDLLFIPGVGHFGEGMRRLRENDLIDFVRKHVEDERYVVGVCLGMQLLFEESEEAPGVKGLSLIEGNVVKLRSRRLPHMGWNEVIFKDTFPNGYYYFVHTYRAVCEEEHVLGTTEYDGEIFPSAVRKGRILGFQFHPEKSSKIGRKLLEKVIECSLSRR'
>>> aligner = Align.PairwiseAligner()
>>> substitution_matrix = dict(blosum62)
>>> for key in substitution_matrix:
...     substitution_matrix[key] += 2 * aligner.epsilon
... 
>>> aligner.substitution_matrix = substitution_matrix
>>> alignments = aligner.align(str1, str2)
>>> for al in alignments:
...     print(al)
... 
ISLKYTAARLHEKGVLLEIEDLQVNQFKNVIFEI--SPTEEVGDFEVKAKFMGVQMETFMLHYQDLLQLQYEGVAVMKLFDRAKVNVNLLIFLLNKKFYGK....................................................................................................
........................XX|XX|XX|X--||XXXXX|X............................................................................................................................................................
MRIGIISVGPGNIMNLYRGVKRASENFEDVSIELVESPRNDLYDLLFIPGVGHFGEGMRRLRENDLIDFVRKHVEDERYVVGVCLGMQLLFEESEEAPGVKGLSLIEGNVVKLRSRRLPHMGWNEVIFKDTFPNGYYYFVHTYRAVCEEEHVLGTTEYDGEIFPSAVRKGRILGFQFHPEKSSKIGRKLLEKVIECSLSRR

**Comment by tanhevg** [comment]
The question is whether BioPython returns all the alignments with maximum score or not. Gotoh algorithm returns them all, so I think this implementation should do that as well. This means that the stopping condition from Durbin et al cited above would only be applicable to zero gap penalties.

**Comment by mdehoon** [comment]
@tanhevg  Why do you think that the Gotoh algorithm returns all alignments with maximum score?

**Comment by MarkusPiotrowski** [comment]
We are not talking about Gotoh, but about Smith-Waterman. AFAIK, in local alignments, alignments always start with a positive score.

**Comment by tanhevg** [comment]
> Why do you think that the Gotoh algorithm returns all alignments with maximum score?

Although the original paper by (Gotoh, 1982) describes the algorithm that returns just a single alignment, extensions that return all alignments have been published a while ago, for example, (Altschul & Erickson, 1986). These papers are quite antiquated; when I was writing my own implementation I was using a very good explanation from a book by (Compeau and Pevzner, 2015). The BioPython implementation returns multiple alignments (not one), so I assumed it returns them all. I don't have the Durbin book at hand to check the details.

> We are not talking about Gotoh, but about Smith-Waterman.

There is a property in `PairwiseAligner` object, that shows the algorithm that is being used. `PairwiseAligner.algorithm` prints "Gotoh local alignment algorithm".

There are actually a few other occasions when alignments with optimal score are skipped (e.g. the one below). Should I raise those as separate bugs? 

###### References
* Gotoh, O. (1982). An improved algorithm for matching biological sequences. Journal of Molecular Biology, 162(3), 705–708. http://doi.org/10.1016/0022-2836(82)90398-9
* Altschul, S. F., & Erickson, B. W. (1986). Optimal sequence alignment using affine gap costs. Bulletin of Mathematical Biology, 48(5–6), 603–616. http://doi.org/10.1007/BF02462326
* Compeau, P., & Pevzner, P. (2015). Bioinformatics Algorithms (2nd ed.). Active Learning Publishers. https://isbnsearch.org/isbn/9780990374619

###### Example of missed asignment
```python 
from Bio import Align
from Bio.SubsMat.MatrixInfo import blosum62

aligner = Align.PairwiseAligner()
aligner.substitution_matrix = blosum62
aligner.mode = 'local'
aligner.open_gap_score = -11
aligner.extend_gap_score = -1

str1 = 'QVLDVMFKGEIFNRNHRSPVATPTPTMSLPEDSKQMVYFAISDYAFNIASRVYHQAGYLNFSITDDMLPHDSGIRLNTKAFRPFTPQIYKKYPDMKLELLGTVVSAPILNVSPGNLSLAPQMEIEGFVILPTSAREPVFRLGVVTNVFASLTFNNSKVTGMLHPDKAQVRLIESKVGMFNVNLFQAFLNYYLLNSLYPDVNAELAQGFPLPLPRHIQLHDLDFQIRKDFLYLGANVQYMR'
str2 = 'KSRDAIQVYTHATSDTMEGLYEQVDKYLEQGYQHIRCQLGFYGGVPENIQTAQNPTQGSYYDQDQYIENTVEMFKNLREKYGKQFHILHDVHERLFPNQAIQFAKQIEQYNPFFIEDILPPSQTEWLDNIRNQSSVSLALGELFNNPEEWKALIINRRVDFIRCHVSQIGGITPALKLGHFCESFGVRIAWHCPPDMTPIGAAVNTHLNVHLHNAAIQEHVEYKANTQRVFPNAAEPINGYLYASEIAGIGVEMDREAAQDFPVEYRPHEWTQSRLPDGSIHTP'

alignments = aligner.align(str1, str2)
for al in alignments:
    print(al.score, al, sep='\n')
print(aligner.algorithm)

# My own implementation
from alignments import SmithWatermanAligner
aligner = SmithWatermanAligner(11, 1, blosum62)
alignment = aligner.align(str1, str2)
print(alignment.score, alignment, sep='\n')
```
Output:
```
31.0
QVLDVMFKGEIFNRNHRSPVATPTPTMSLPEDSKQMVYFAISDYAFNIASRVYHQAGYLNFSITDDMLPHDSGIRLNTKAFRPFTPQIYKKYPDMKLELLGTVVSAPILNVSPGNLSLAPQMEIEGFVILPTSAREPVFRLGVVTNVFASLTFNNSKVTGMLHPDKAQVRLIESKVGMFNVNLFQAFLNYYLLNSLYPDVNAELAQGFPLPLPRHIQLHDLDFQIRKDFLYLGANVQYMR............................................................................................................................................................................................................................
................................................................................................................................................................................................XXX||XX|XXX|X||X............................................................................................................................................................................................................................................................
................................................................................................................................................................................KSRDAIQVYTHATSDTMEGLYEQVDKYLEQGYQHIRCQLGFYGGVPENIQTAQNPTQGSYYDQDQYIENTVEMFKNLREKYGKQFHILHDVHERLFPNQAIQFAKQIEQYNPFFIEDILPPSQTEWLDNIRNQSSVSLALGELFNNPEEWKALIINRRVDFIRCHVSQIGGITPALKLGHFCESFGVRIAWHCPPDMTPIGAAVNTHLNVHLHNAAIQEHVEYKANTQRVFPNAAEPINGYLYASEIAGIGVEMDREAAQDFPVEYRPHEWTQSRLPDGSIHTP

31.0
......................................................QVLDVMFKGEIFNRNHRSPVATPTPTMSLPEDSKQMVYFAISDYAFNIASRVYHQAGYLNFSITDDMLPHDSGIRLNTKAFRPFTPQIYKKYPDMKLELLGTVVSAPILNVSPGNLSLAPQMEIEGFVILPTSAREPVFRLGVVTNVFASLTFNNSKVTGMLHPDKAQVRLIESKVGMFNVNLFQAFLNYYLLNSLYPDVNAELAQGFPLPLPRHIQLHDLDFQIRKDFLYLGANVQYMR
............................................................................................................................................................................................................................................................XXXX|X||X||XXXXX|.........................
KSRDAIQVYTHATSDTMEGLYEQVDKYLEQGYQHIRCQLGFYGGVPENIQTAQNPTQGSYYDQDQYIENTVEMFKNLREKYGKQFHILHDVHERLFPNQAIQFAKQIEQYNPFFIEDILPPSQTEWLDNIRNQSSVSLALGELFNNPEEWKALIINRRVDFIRCHVSQIGGITPALKLGHFCESFGVRIAWHCPPDMTPIGAAVNTHLNVHLHNAAIQEHVEYKANTQRVFPNAAEPINGYLYASEIAGIGVEMDREAAQDFPVEYRPHEWTQSRLPDGSIHTP..........

31.0
..................................................QVLDVMFKGEIFNRNHRSPVATPTPTMSLPEDSKQMVYFAISDYAFNIASRVYHQAGYLNFSITDDMLPHDSGIRLNTKAFRPFTPQIYKKYPDMKLELLGTVVSAPILNVSPGNLSLAPQMEIEGFVILPTSAREPVFRLGVVTNVFASLTFNNSKVTGMLHPDKAQVRLIESKVGMFNVNLFQAFLNYYLLNSLYPDVNAEL----AQGFPLPLPRHIQLHDLDFQIRKDFLYLGANVQYMR
.............................................................................................................................................................................................................................................X|X||XX|XXXXXXX|X----||X||XXXXX|.........................
KSRDAIQVYTHATSDTMEGLYEQVDKYLEQGYQHIRCQLGFYGGVPENIQTAQNPTQGSYYDQDQYIENTVEMFKNLREKYGKQFHILHDVHERLFPNQAIQFAKQIEQYNPFFIEDILPPSQTEWLDNIRNQSSVSLALGELFNNPEEWKALIINRRVDFIRCHVSQIGGITPALKLGHFCESFGVRIAWHCPPDMTPIGAAVNTHLNVHLHNAAIQEHVEYKANTQRVFPNAAEPINGYLYASEIAGIGVEMDREAAQDFPVEYRPHEWTQSRLPDGSIHTP..........

Gotoh local alignment algorithm
31.0
LNYYLLNS----LYPDVNAELAQGFPLPLPRH
+|.||..|    +..+++.|.||.||+....|
INGYLYASEIAGIGVEMDREAAQDFPVEYRPH
```

**Comment by MarkusPiotrowski** [comment]
@tanhevg 
It's about a proper implementation of an algorithm. In local alignments according to Smith & Waterman, you fill the score matrix only with positive scores. Negative scores and zero scores are just stored as zero. During backtracing, the alignment stops when it encounters the first 0 (https://www.sciencedirect.com/science/article/pii/0022283681900875). Or, to put it in other words, a local alignment according to Smith & Waterman **will always start with a positive score**. 
The fact, that ```pairwise2``` and the new ```Align``` module and EMBOSS' Water all give the same results is a good indication that we implemented the SW algorithm correctly. 

Regarding your last example, I'm not really sure if this is an error. If I only take the short sequences (LNYY...PRH and INGY...RPH) and do a **local alignment** then I don't get the result that you showed, but I get it when using a **global** alignment. So my guess would be that the early gaps bring the score down to zero, which would be a stop condition during backtrace in SW (but not in Needleman & Wunsch). Hard to say without looking at the matrices, but maybe you can have a look.  

So, I see why you don't like the results but this is the correct behaviour for SW local alignments.

**Comment by MarkusPiotrowski** [comment]
May want to add that "Gotoh" itself is not a different way how to find optimal alignments. But Gotoh showed that, when using affine gap penalties (which is usually done), the calculation for each cell in the matrix is only dependent on the neighbouring cells.

**Comment by tanhevg** [comment]
@MarkusPiotrowski , I think the algorithm that is outlined in the original S&W paper makes some implicit assumptions about the scoring scheme, which do not always hold for the commonly used schemes. Namely, that all scores (match, mismatch, gap open and extend) are non zero. 

Let's consider a simple scenario: align two amino acid sequences, `IIAA` and `CCKK`, using BLOSUM62 substitution matrix, gap opening penalty of -11 and extension penalty of -1 (in BioPython notation). The A-C score in BLOSUM62 is zero, so the correct alignments would be `A-C` and `AA-CC` with zero score. EMBOSS water 6.6.0 incorrectly reports an `I-C` alignment with a zero score (the correct score for this alignment is -1). BioPython does not report any alignments. 

This example demonstrates that EMBOSS cannot always be used as a benchmark.

So I think we first need to agree what is an optimal local alignment. I thought that it is an alignment of any two (non-empty) substrings, possibly with gaps, that attains the maximum score (under the given scheme). Stopping at zero, as in the paper, imposes an additional requirement on the alignment, that it should attain the maximum **positive** score. At the very least all of this needs to be clarified in the docs.
```
$ ./water /tmp/iiaa.txt /tmp/cckk.txt -datafile EBLOSUM62 -gapopen 10 -gapextend 1 -stdout -auto
########################################
# Program: water
# Rundate: Tue  5 Feb 2019 13:51:01
# Commandline: water
#    [-asequence] /tmp/iiaa.txt
#    [-bsequence] /tmp/cckk.txt
#    -datafile EBLOSUM62
#    -gapopen 10
#    -gapextend 1
#    -stdout
#    -auto
# Align_format: srspair
# Report_file: stdout
########################################

#=======================================
#
# Aligned_sequences: 2
# 1: 
# 2: 
# Matrix: EBLOSUM62
# Gap_penalty: 10.0
# Extend_penalty: 1.0
#
# Length: 1
# Identity:       0/1 ( 0.0%)
# Similarity:     0/1 ( 0.0%)
# Gaps:           0/1 ( 0.0%)
# Score: 0.0
# 
#
#=======================================

                   1 I      1
                     .
                   1 C      1


#---------------------------------------
#---------------------------------------
(ampliphyre) et517@ld-mjeste2:~/tools/EMBOSS-6.6.0/emboss
$ cat /tmp/iiaa.txt
IIAA
(ampliphyre) et517@ld-mjeste2:~/tools/EMBOSS-6.6.0/emboss
$ cat /tmp/cckk.txt
CCKK
```

**Comment by MarkusPiotrowski** [comment]
At least Biopython does what I would expect here: reporting no alignment. Don't know why EMBOSS' Water reports an alignment at all, still, the score is the highest (and lowest) it can obtain in this case: 0 (because there are no negative values in a score matrix of SW local alignments).

As I said before, I'm talking about an implementation of the Smith-Waterman algorithm here. For ```pairwise2``` I tried to implement the algorithms as they are described in the literature. I guess the same is true for @mdehoon and the new ```Aligner```.

If you think there should be a different way of doing local alignments you may think about doing a pull request for implementation in Biopython.

I think the 'problem' is maybe due to implicit assumptions about the aims of global vs. local alignments: In a global alignment, you are *comparing* two sequences, so you always want to have a (the maximum) score, even if it is negative. In local alignments, you are *looking/searching for* similar subsequences, if there are none, there is no need to report them.

**Comment by mdehoon** [comment]
@tanhevg 

> Although the original paper by (Gotoh, 1982) describes the algorithm that returns just a single alignment, extensions that return all alignments have been published a while ago, for example, (Altschul & Erickson, 1986). 

I haven't checked the book by Compeau & Pevzner yet, but the articles by Gotoh (1982) and Altschul & Erickson (1986) seem to deal with global alignments only.

In your implementation, how does the algorithm work? There must be some difference between your implementation and the description of local alignment algorithms by Richard Durbin et al.

**Comment by tanhevg** [comment]
@mdehoon , My implementation is not fully finished and tested yet. In keeping with the approach from Compeau & Pevzner, 2015, I handle the local alignments very similarly to global alignments. The main difference is that for local alignments it is possible to get from source cell `(0,0)` of the matrix to any cell `(i, j)` without paying a penalty, and similarly from any cell `(i, j)` to the sink cell `(n, m)`. These edges with zero cost are called the free rides. When backtracking, I am tracking those free rides in addition to edges between neighbouring cells.

**Comment by MarkusPiotrowski** [comment]
I don't have the book but today I saw some slides and videos from Compeau & Pevzner and I can't see that they do something different for local alignments than the scoring scheme of Smith & Waterman. They just use other words. The "free ride" with zero cost means that each cell in the score matrix will have a non-negative score. S(i,j) = max{0 (free ride), S(i-1, j-1) + match/mismatch, S(i-1, j) + gap, S(i, j-1) + gap). So it's either zero or positive. 

It would really be interesting to see your actual implementation. Above you showed an example that it works in principal.

**Comment by mdehoon** [comment]
@tanhevg  @MarkusPiotrowski  It is possible to get all alignments by keeping track in the trace matrix of the cells that can act as the starting point of the alignment. But I still wonder if this is a good thing. In the original example, you would get 3 possible starting points, and 2 possible end points, so 6 alignments for what is essentially the same thing. If I had a choice, I would take the shortest alignment only, as it gives the highest score per amino acid, and seems most consistent with the purpose of doing a local alignment.

I also checked the book by Compeau and Pevzner. I didn't see a difference in their local alignment algorithm compared to the one described by Richard Durbin et al., and as far as I can tell they don't insist that all possible alignments should be returned.

I guess then that the choice is up to us. To me, reasonable solutions are either (a) keep the current implementation, or (b) modify the algorithm such that zero-score extensions are not included on either side of the alignment. Option (b) has the advantage that the left- and right-side of the alignment are handled consistently.

**Comment by MarkusPiotrowski** [comment]
@tanhevg  @mdehoon Although I also acknowledge that option (b) is more consistent how to handle 'zero-extensions' on both sides of the alignment, it makes backtracing more difficult (check each cell with maximum score if it is the result of a 'zero-extension'). Also, I feel that this is not part of the original algorithm. 

So I would vote for option (a).

I wonder if the reason that EMBOSS Water returns the shorter alignment in the above example only, is just by pure 'chance', since it always returns only one alignment (maybe it always returns that alignment that results from backtracing from the earliest cell with maximum score).

**Comment by mdehoon** [comment]
@MarkusPiotrowski  It wouldn't make the algorithm much more difficult; all you need to do when back-tracing is to discard an alignment if you find an alternative end point.

**Comment by mdehoon** [comment]
@MarkusPiotrowski  In 1000 alignments by EMBOSS water of randomly generated sequences with multiple ending points for the same starting points, in all cases the shortest alignment was returned.

**Comment by MarkusPiotrowski** [comment]
@mdehoon My suspiscion it, since it EMBOSS Water only returns one alignment, it stores the position of the cell with the highest score already during the score table calculation and replaces the value only if there is a higher score (e.g. ```if actual_score > max_score: traceback_start = (i,j) ...```). Thus the backtrace would always start at the earliest possible cell (most up, most left) and you have no zero-value extensions. When I wrote 'by chance' I meant that we don't know if it was the explicit intention of the programmer to disallow zero-extensions or if it is 'just' a consequence of how the backtrace start position is determined.

Interestingly, the implementation of ```lalign``` in EMBOSS returns alignments **with leading zero extensions** (but not trailing zero extensions) ....

However, in Waterman & Eggert (1987) the authors confirm that by using Smith&Waterman "sometimes alignments can have 0 score segments added on to them". And they decide to dissallow this! ("We simply do not allow these, as is described below"). This makes me now agreeing to your option (b), i.e. to change our implementation. 
I recommend to make this clear in our documentation: in the docstrings as well as in the tutorial.

**Comment by MarkusPiotrowski** [comment]
@mdehoon 
Do you agree:
-A local alignment never starts or ends with a gap
-Thus from a cell with a maximum score the trace back must go diagonally
-So if the cells S(i,j) and S(i-1, j-1) have the same maximum score, then S(i,j) cannot be a valid starting point for the traceback

**Comment by mdehoon** [comment]
@MarkusPiotrowski  Yes I agree. But that may not be sufficient to avoid all zero-score extensions.

**Comment by mdehoon** [comment]
Fixed for the PairwiseAligner in Bio.Align by #1938  by disallowing zero-score extensions.

**Comment by MarkusPiotrowski** [comment]
@mdehoon and I agreed that zero-score extensions should not been added to local alignments. Biopython's two pairwise aligners ```pairwise2``` and ```PairwiseAligner``` up to now prevented *leading* zero-score extensions but not *trailing* zero-score extensions.
This has been addressed now with PRs #1938 and #1955. 

Closing this now.
